fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,496 @@
+                if (cell_ok)
+                    break;
+
+                if (driver_in_port.second.type != PORT_IN)
+                    continue;
+
+                CellInfo *driver_cell_driver = driver_in_port.second.net->driver.cell;
+
+                for (auto &cell : cluster_driver_in_out_cells.at(cluster)) {
+                    if (IdString(cell) == driver_cell->type && driver_cell_driver == next_cell) {
+                        cell_ok = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!cell_ok or driver_cell == nullptr)
+                continue;
+
+            BelRange bels_in_tile = getBelsByTile(next_loc.x, next_loc.y);
+            for (auto bel : bels_in_tile) {
+                auto &bel_data = bel_info(ctx->chip_info, bel);
+
+                if (bel_data.site != next_bel_data.site)
+                    continue;
+
+                IdString bel_name(bel_data.name);
+
+                auto res_placed = std::find(placed_bels.begin(), placed_bels.end(), bel_name);
+                auto res_bels = std::find(port.second.begin(), port.second.end(), bel_name);
+                bool bel_valid = ctx->isValidBelForCellType(driver_cell->type, bel);
+
+                if (!bel_valid || res_bels == port.second.end() || res_placed != placed_bels.end())
+                    continue;
+
+                //log_info("\tNet name: %s - Cell name: %s - Cell type: %s - Bel: %s\n", next_cell->ports.at(driver_port).net->name.c_str(ctx), driver_cell->name.c_str(ctx), driver_cell->type.c_str(ctx), ctx->nameOfBel(bel));
+                placement.push_back(std::make_pair(driver_cell, bel));
+                placed_bels.push_back(bel_name);
+
+                break;
+            }
+        }
+
+        // Get pattern to find the next cell in the chain
+        auto cell_pattern = cell_pattern_map.find(next_cell->name);
+        std::pair<IdString, IdString> pattern_cfg = cell_pattern->second;
+        IdString port = pattern_cfg.first;
+        if (port == IdString()) {
+            log_error("Cluster misconfiguration. None of the patterns match the '%s' cell\n", next_cell->type.c_str(ctx));
+        }
+
+        next_net = next_cell->ports[port].net;
+
+        // Check if we have reached the end of the chain
+        if (next_net == nullptr || next_net->users.size() == 0) {
+            break;
+        }
+
+        next_cell = next_net->users.at(0).cell;
+
+        next_bel = find_next_chained_bel(ctx, next_bel, ctx->id("CO3"));
+        NPNR_ASSERT(next_bel != BelId());
+
+        next_loc = ctx->getBelLocation(next_bel);
+
+    } while(true);
+
+    return true;
+}
+
+ArcBounds Arch::getClusterBounds(ClusterId cluster) const
+{
+    NPNR_ASSERT(cluster != ClusterId());
+    CellInfo *root_cell = getClusterRootCell(cluster);
+    CellInfo *last_cell = packed_clusters.at(cluster).back();
+    Loc loc_root = getBelLocation(root_cell->bel);
```

Comment:
these functions (this and `getClusterOffset`) are only intended to return approximations, but it's important that they function **before** the chain has been placed (that is in fact their main use, after the chain has been placed anything downstream that wants to know approximate bounds/offsets of the chain can just query actual positions). So this current implementation isn't really the intended purpose of these functions.

Reply:
I might have missed something, but I have noticed that the `getClusterBounds` is not called at the moment in any of the algos. At the moment I left it blank, with a  `TODO`, to fill it as soon as it will be necessary.

I have also implemented a "fixed" heuristic for the clusters that span multiple tiles, hacked for the moment to return a valid offset for the carry-chains. I think that this kind of information could be provided as a device configuration in the cluster info.

Reply:
yeah, there are some possible future placer ideas I had that meant it could be useful, but a stub is alright for now `getClusterOffset` is the more important one for HeAP.